### PR TITLE
WIP Application state

### PR DESCRIPTION
**Pull request by jasongrout**
Here are some thoughts about the states in the application.

### CODE REVIEW

File: src/state/state.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+---------------
+- What documents are open (i.e., what document sessions we are talking with)
+
+FileBrowser
+-----------
+- Currently open directory in file browser
+
+Individual documents
+--------------------
+- ephemeral editing state, like code folding, etc?
```
Comment (blink1073) [inline src/state/state.md:30]: I was about to say that this is a step too far, but SublimeText actually does preserve cursor position on a quit/relaunch.

File: src/state/state.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
+We have several different notions of state, with different lifecycles on each.
+
+The Lab has a number of open "documents".  A document is represented by a top-level widget in the dock area (but possibly not all top-level widgets are documents - we could tear off a part of a document to make a top-level widget in the dock area that is a part of a document...).  Examples of documents include any file that is represented, as well as terminals, chat windows, real-time video conferences, etc.
+
+Lab IDE Session state
+=====================
+
+Plugins should be able to read/write slices of the IDE state. This IDE state should be preserved on refresh, so a Lab IDE session id should be stored in the URL and the state periodically synced to the server.
+
+Initial implementation could be a local storage solution, tied to the URL.  When the lab starts up, it randomly appends a session id, and then stores session information in local storage tied to the url.  In general, the lab would ask the server for a session id if one is not provided?
```
Comment (blink1073) [inline src/state/state.md:10]: I think so, that is how notebook sessions are handled.

File: src/state/state.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+Question: on a refresh, should Lab IDE session state only inform the initial setup of the environment (and then the URL changes to reflect a new lab sesion id), or should we try to keep two lab sessions in sync, so if you change the dock layout in one browser window, the dock layout in the other browser changes automatically? Or should we make it easy for a user to 'clone' a lab session?
```
Comment (blink1073) [inline src/state/state.md:12]: I'd say the state sharing settings would by default not share things like dock layout, but provide the ability to share more/clone more.

Comment (jasongrout) [inline src/state/state.md:12]: I was thinking that the lab sessions would be tied to the user, i.e., the authentication system would prevent you from going to my lab session URL and seeing my session.

File: src/state/state.md (hunk 1 above)
Comment (jasongrout) [inline src/state/state.md:30]: Interesting.  I believe that google docs does not preserve cursor position (which wouldn't really make sense, since I don't think it has the concept of a user session that spans refreshes).

**Comment by bollwyvl** [comment]
For actually remembering the state in the browser, I am a huge proponent of [baobabjs](https://github.com/Yomguithereal/baobab), which was inspired by zippers from closure and cursors from Om.
- First you make a tree. It holds an object.
  - you can put any JSON object in there
  - you can listen to very specific parts of the tree with a _cursor_
    - which can then drive gross data changes and draw UI
  - inside user behavior, you can write _back_ to any cursor
    - and validate that the change worked
      - optionally allowing failed state to be stored or rejected
  - you can create computed data (monkeys) that can also be listened to as a cursor

Here's an ES5 functional strawman, but it works well with ES6 and typescript classes as well.

``` js
var opts = {
  validate: function(previousState, newState, affectedPaths){
    // maybe apply JSON schema?
    // return new ValidationError()
  } 
};

var tree = new Baobab({
  state: localStorage.getItem("state") || {},
  notebooks: {Untitled1: {metadata:{}, cells:[]}},
  _specialCells: monkey(
    [".", "state"],
    [".", "notebooks"],
    function(state, notebooks) {
      // perhaps do something fancy like figure out if they are on the screen
      return Object.keys(notebooks)
        .reduce(function(memo, key){
          return memo.concat(notebooks[key].cells);
        }, []);
    });
}, opts);

var cursor = {
  root: tree.root,
  state: tree.select(["state"]),
  notebooks: tree.select(["notebooks"]),
  specialCells: tree.select(["_specialCells"])
};

cursor.root.on("update", function(){
  /* do something when any _data_ changes */
});

cursor.state.on("update", function(){
  /* do something when just state changes */
  localStorage.setItem("state", cursor.state.get()),
});

cursor.notebooks.on("update", function(){
  /* do something when just notebooks change */  
});

cursor.specialCells.on("update", function(){
  /* do something when just the special cells change */
});
```

As for structuring the data, I'll always, always throw [JSON-LD](http://json-ld.org) out there first. It already supports:
- strong IDs as URI/IRIS
- namespaces
- internationalization
- sane imports
- different operations: expansion, compaction, flattening, framing, normalization, signing

Basically, you'd have a canonical, collections-in-documents-style (i.e. one per widget or something) format, and give just about everything IDs. then, you'd be able to project the data from the canonical form into widget-specific, useful views.

Combining both of these would be a very powerful API that was basically self-documenting, making it very easy to hook in at the power user level with TypedCursorFactoryProviders as well as more mundane functional approaches, a la riot.js or mithril, which might have less of a learning cliff than a full future-forward stack.

**Comment by jasongrout** [comment]
@bollwyvl - thanks for the pointers! I've been looking into what might be a good library to use to manage the state.

Just curious: what do you think of om.next, https://github.com/omcljs/om/wiki/Quick-Start-(om.next)? (I think it's probably not a good fit for us, but it proposes some interesting ideas to deal with problems that come up with cursors, etc.)

**Comment by jasongrout** [comment]
(notes to self) Here are a few more articles about baobab vs react, etc.:
- http://www.christianalfoni.com/articles/2015_11_16_Flux-vs-Single-State-Tree
- [immstruct](https://github.com/omniscientjs/immstruct) - cursors built around immutable.js

**Comment by minrk** [comment]
:+1: to immutable.js

**Comment by jasongrout** [comment]
We are now working on managing individual document state as part of plugins, rather than application-level state.
